File: content/snippets/scala3/design_patterns_strategy.scala
// Modern Scala Strategy using function values and Scala 3 type classes (given/using)
// No Strategy trait or subclasses - strategies are plain function values.

// Approach 1: function value (simplest)
class Calculator(val strategy: (Int, Int) => Int):
  def execute(a: Int, b: Int): Int = strategy(a, b)

// Approach 2: type class with given/using (Scala 3) - zero-overhead, compile-time resolved
trait Strategy:
  def execute(a: Int, b: Int): Int

given addStrategy: Strategy with
  def execute(a: Int, b: Int) = a + b

given subtractStrategy: Strategy with
  def execute(a: Int, b: Int) = a - b

def calculate(a: Int, b: Int)(using s: Strategy): Int = s.execute(a, b)

@main def strategyModernExample(): Unit =
  // Function value approach
  val add  = Calculator(_ + _)
  val sub  = Calculator(_ - _)
  val mul  = Calculator(_ * _)
  println(s"Add: ${add.execute(5, 3)}")
  println(s"Subtract: ${sub.execute(5, 3)}")
  println(s"Multiply: ${mul.execute(5, 3)}")

  // Type class approach
  println(s"Given add: ${calculate(5, 3)(using addStrategy)}")
  println(s"Given sub: ${calculate(5, 3)(using subtractStrategy)}")

File: content/snippets/python/design_patterns_strategy.py
from abc import ABC, abstractmethod

class Strategy(ABC):
    @abstractmethod
    def execute(self, a: int, b: int) -> int:
        pass

class AddStrategy(Strategy):
    def execute(self, a: int, b: int) -> int:
        return a + b

class SubtractStrategy(Strategy):
    def execute(self, a: int, b: int) -> int:
        return a - b

class MultiplyStrategy(Strategy):
    def execute(self, a: int, b: int) -> int:
        return a * b

class Context:
    def __init__(self):
        self.strategy: Strategy = None
    
    def set_strategy(self, strategy: Strategy) -> None:
        self.strategy = strategy
    
    def execute_strategy(self, a: int, b: int) -> int:
        return self.strategy.execute(a, b)

ctx = Context()

ctx.set_strategy(AddStrategy())
print(f"Add: {ctx.execute_strategy(5, 3)}")

ctx.set_strategy(SubtractStrategy())
print(f"Subtract: {ctx.execute_strategy(5, 3)}")

ctx.set_strategy(MultiplyStrategy())
print(f"Multiply: {ctx.execute_strategy(5, 3)}")
